app/celery/nightly_tasks.py

```diff
@@ -334,3 +335,42 @@ def save_daily_notification_processing_time(bst_date=None):
             messages_within_10_secs=result.messages_within_10_secs,
         )
     )
+
+
+@notify_celery.task(name="delete-oldest-quarter-of-unneeded-notification-history")
+def delete_oldest_quarter_of_unneeded_notification_history():
+    # This retention limit is hardcoded for the moment. It is picked from
+    # https://github.com/alphagov/notifications-aws/blob/main/decisions/2022-12-01-notification-history-retention-period.md
+    # and can be increased in the future when we move 3 months into the next financial year
+    retention_limit = datetime(2022, 4, 1)
+
+    oldest_record = db.session.query(func.min(NotificationHistory.created_at)).one()[0]
+    if oldest_record >= retention_limit:
+        current_app.logger.info(
+            "No notifications older than retention date of %s so nothing to delete", retention_limit
+        )
+        return
+
+    # Pick a potential deletion date to start, that we know is older than the oldest notification
+    # in notification_history at the time of writing this code
+    deletion_date = datetime(2019, 10, 1)
```

Comment:
The `deletion_date` must be the start of quarter otherwise we risk deleting data that is still within the retention limit. Should we do a check somewhere in the `while` loop to check that `deletion_date` is earlier than the `retention_limit`? 

It's not a problem with the code as it is now, it would be more to stop someone accidentally changing `deletion_date` to `datetime(2022, 2, 1)` (for example) and then deleting everything up to 1st May 2022 instead of 1st April 2022.

Reply:
Good suggestion.

I'll add a commit with this. Not only will it cover the case where someone decides to change the `deletion_date`, but also the case where there is some type of other bug in the while loop logic.